flutter/tonic/dart_isolate_reloader.cc

```diff
@@ -0,0 +1,330 @@
+    Dart_Handle result = Dart_Null();
+    switch (tag_) {
+      case Dart_kImportTag:
+        result = Dart_LoadLibrary(uri, source, 0, 0);
+        break;
+      case Dart_kSourceTag:
+        result = Dart_LoadSource(library, uri, source, 0, 0);
+        break;
+      case Dart_kScriptTag:
+        result = Dart_LoadScript(uri, source, 0, 0);
+        break;
+    }
+
+    if (Dart_IsError(result)) {
+      return result;
+    }
+    return Dart_Null();
+  }
+
+ private:
+  bool success_;
+  intptr_t tag_;
+  std::string url_;
+  std::string library_url_;
+  std::string error_;
+  std::vector<uint8_t> payload_;
+};
+
+DartIsolateReloader::DartIsolateReloader(DartLibraryProvider* library_provider)
+    : thread_(new base::Thread("DartIsolateReloader")),
+      library_provider_(library_provider),
+      monitor_(),
+      load_error_(Dart_Null()),
+      pending_requests_(0) {
+  CHECK(thread_->Start());
+}
+
+DartIsolateReloader::~DartIsolateReloader() {
+}
+
+void DartIsolateReloader::SendRequest(Dart_LibraryTag tag,
+                                      Dart_Handle url,
+                                      Dart_Handle library_url) {
+  scoped_refptr<base::TaskRunner> runner =
+      thread_->message_loop()->task_runner();
+
+  MonitorLocker ml(&monitor_);
+
+  // Post a task to the worker thread. This task will request the I/O and
+  // post a LoadResult to be processed once complete.
+  if (runner->PostTask(
+        FROM_HERE,
+        base::Bind(&DartIsolateReloader::RequestTask,
+                   library_provider_,
+                   this,
+                   static_cast<intptr_t>(tag),
+                   StdStringFromDart(url),
+                   StdStringFromDart(library_url)))) {
+    pending_requests_++;
+  }
+}
+
+void DartIsolateReloader::PostResult(std::unique_ptr<LoadResult> load_result) {
+  MonitorLocker ml(&monitor_);
+  pending_requests_--;
+  load_results_.push(std::move(load_result));
+  ml.Notify();
+}
+
+// As each source file is requested, a LoadRequest is queued to be processed on
+// worker thread.
+class DartIsolateReloader::LoadRequest : public DataPipeDrainer::Client {
+ public:
+  LoadRequest(DartLibraryProvider* library_provider,
+              DartIsolateReloader* isolate_reloader,
+              intptr_t tag,
+              const std::string& url,
+              const std::string& library_url)
+      : isolate_reloader_(isolate_reloader),
+        tag_(tag),
+        url_(url),
+        library_url_(library_url),
+        weak_factory_(this) {
+    library_provider->GetLibraryAsStream(
+        url_,
+        base::Bind(&LoadRequest::OnStreamAvailable,
+                   weak_factory_.GetWeakPtr()));
+  }
+
+ protected:
+  void OnStreamAvailable(mojo::ScopedDataPipeConsumerHandle pipe) {
+    if (!pipe.is_valid()) {
+      std::unique_ptr<DartIsolateReloader::LoadResult> result =
+          std::unique_ptr<DartIsolateReloader::LoadResult>(
```

Comment:
std::unique_ptr has a cosntructor that makes thsi less verbose.  Rather than using `=` to initialize the local, you can just call the constructor with `(`